packages/block-library/src/query/view.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * WordPress dependencies
+ */
+import { store, navigate, prefetch } from '@wordpress/interactivity';
+
+const isValidLink = ( ref ) =>
+	ref &&
+	ref instanceof window.HTMLAnchorElement &&
+	ref.href &&
+	( ! ref.target || ref.target === '_self' ) &&
+	ref.origin === window.location.origin;
+
+const isValidEvent = ( event ) =>
+	event.button === 0 && // left clicks only
+	! event.metaKey && // open in new tab (mac)
+	! event.ctrlKey && // open in new tab (windows)
+	! event.altKey && // download
+	! event.shiftKey &&
+	! event.defaultPrevented;
+
+store( {
+	selectors: {
+		core: {
+			query: {
+				startAnimation: ( { context } ) =>
+					context.core.query.animation === 'start',
+				finishAnimation: ( { context } ) =>
+					context.core.query.animation === 'finish',
+			},
+		},
+	},
+	actions: {
+		core: {
+			query: {
+				navigate: async ( { event, ref, context, state } ) => {
+					if ( isValidLink( ref ) && isValidEvent( event ) ) {
+						event.preventDefault();
+
+						const id = ref.closest( '[data-wp-navigation-id]' )
+							.dataset.wpNavigationId;
+
+						// Don't announce the navigation immediately, wait 300 ms.
+						const timeout = setTimeout( () => {
+							context.core.query.message =
+								state.core.query.loadingText;
+							context.core.query.animation = 'start';
+						}, 300 );
+
+						await navigate( ref.href );
+
+						// Dismiss loading message if it hasn't been added yet.
+						clearTimeout( timeout );
+
+						// Announce that the page has been loaded. If the message is the
+						// same, we use a no-break space similar to the @wordpress/a11y
+						// package: https://github.com/WordPress/gutenberg/blob/c395242b8e6ee20f8b06c199e4fc2920d7018af1/packages/a11y/src/filter-message.js#L20-L26
+						context.core.query.message =
+							state.core.query.loadedText +
+							( context.core.query.message ===
+							state.core.query.loadedText
+								? '\u00A0'
+								: '' );
+
+						context.core.query.animation = 'finish';
+
+						// Focus the first anchor of the Query block.
+						document
+							.querySelector(
+								`[data-wp-navigation-id=${ id }] a[href]`
+							)
+							?.focus();
```

Comment:
We should manage the focus to return to the clicked button so we're not stealing focus to a new place. The `current-page` element can remain an `<a>` element so it can receive focus and simplify the code. I'm not 100% positive of this interaction though. I think it would be better than always returning to the first item though.

Reply:
We will address this in a follow-up PR. 👍

Reply:
Hi, @jeryj. I was about to open a PR to address these changes. 🙂 

You propose to keep a link to the current page with `aria-current=page`, and maintain focus there, am I right? There is a case I don't know how to manage: when the Pagination Number block is not present. What would be the best approach in this case?

Reply:
This is a tricky one because focus probably needs to go back to the newly loaded results. One could make an argument that this makes the experience worse because you could not keep hitting next but you also can't do this when the page reloads either.

Reply:
Moving focus to the results after load likely simplifies the code and makes the focus experience more consistent, so sounds good to me! Thanks!

Reply:
Again, thanks for your feedback, @jeryj and @alexstine. 😊

This is also the current implementation, so no more changes are needed. Setting this conversation as resolved.